### ISSUE TITLE
MatSelect accessibility rework

### ISSUE THREAD

**Issue by jelbourn** [angular/components#11083]
This issue tracks rework of the MatSelect component to resolve multiple accessibility issues. Based on [this prototype](https://stackblitz.com/edit/angular-a8j36n?file=app%2Fapp.component.html) (just the a11y behavior and API, not the visuals)

Summary:
* New template-based syntax for options driven by data instead of ngFor (the existing syntax will remain for backwards compatibility for at least two major versions)
* Option text will always be in the DOM, but not the option _templates_
* Interaction changes from `listbox` to `combobox`

This should capture issues reported in #5220, #10292, #6382, #3587, #1670

**Comment by jasonzhang2022** [comment]
mat-paginator uses mat-selector for the number of items of per page. So mat-paginator can't be used by VoiceOver, too.

**Comment by cec** [comment]
@jelbourn I hope I'm not wrong, but mat-select is used by mat-autocomplete as well right?

If so, will this rework would add the possibility for paginated autocompete suggestions as discussed in #5046 ?

Thank you!

**Comment by jelbourn** [comment]
They're separate components but do share some code; that feature looks to be orthogonal to this.

**Comment by cec** [comment]
Hi @jelbourn , please pardon my ignorance, by orthogonal do you mean unrelated to this rework?

**Comment by jelbourn** [comment]
Yeah, I mean this rework doesn't affect that feature one way or another.

**Comment by JamieMcI** [comment]
Are there any known workarounds to this issue while waiting on the rework? I'm thinking of trying to style an HTML Select to look similar with a floating label, etc.

**Comment by jelbourn** [comment]
Using a native select will certainly give better a11y.

**Comment by benelliott** [comment]
@jelbourn Do you think there would be any value in moving the core and accessibility logic of MatSelect into a new CdkSelect component, so that MatSelect only provides the Material-specific styling on top of this? Creating a component that is functionally equivalent to native `select` is not at all trivial so it could be a very useful resource for people who want to make custom comboboxes without using Material Design.

**Comment by jelbourn** [comment]
@benelliott  It's something I'd like to do; we're still figuring out what we want the final API to look like. Once that's settled, we can figure out which parts make sense to be in the cdk.

**Comment by jasonzhang2022** [comment]
Any update on this one?

**Comment by jrood** [comment]
@jelbourn If these discussions about the API need to remain internal, that's cool. Otherwise I'd be happy to contribute if you can point me to a particular thread.

**Comment by tinayuangao** [comment]
We decided to keep the current API while improve the accessibility. I will send out a PR soon

**Comment by jrood** [comment]
Thanks @tinayuangao ! Do you know if the a11y changes are such that https://github.com/angular/material2/pull/7835 should be kept on hold, or if keeping the current API means that https://github.com/angular/material2/pull/7835 can be merged?

**Comment by kammokiran** [comment]
Hi, I have an application where the mat-select has been used. It works fine with keyboard and the screen reader (jaws 17 and 18). I am able to select an option inside the select box. However once I come out of the select box, Jaws doesn't announce the selected value. Please could you suggest a sollution for this issue.

**Comment by kammokiran** [comment]
[](https://material.angular.io/components/select/overview)

**Comment by divyaadurai** [comment]
I am still facing the issue on reading mat-option text while doing up-arrow/down-arrow with JAWS 2018.1807.8 
-  Browsers which doesn't support are chrome 68, Edge 42, Firefox 61.0.1.
-  Supported browser IE 11.1 (reading properly the text in mat-option)
Windows Version 1803

Is there any fix for this issue available ?
@kammokiran: Is the option text is reading properly. Any suggestions on updating browser versions or anything else ?

**Comment by kammokiran** [comment]
@divyaadurai, try using alt key + the down arrow key first, then press the down and up arrow keys to browse through the options inside the drop down. I do not encounter issues while browsing the options, I just press either the space or enter key to get into the drop down and use the arrow keys there on.

**Comment by Splaktar** [comment]
Adding my findings to this as well. The [mat-select Examples page](https://material.angular.io/components/select/examples) scores a `66` out of `100` on the Lighthouse audit.

![screen shot 2018-10-05 at 5 41 58 pm](https://user-images.githubusercontent.com/3506071/46561521-094e3400-c8c6-11e8-9e29-f5978fc2d9d2.png)

This is due to the problem mentioned above where the `role="option"` elements are not in the DOM until the dropdown panel is opened. This isn't valid according to [Axe](https://dequeuniversity.com/rules/axe/2.2/aria-required-children?application=lighthouse) and the [WAI-ARIA 1.0 Listbox Spec](https://www.w3.org/TR/wai-aria/#listbox).

This issue means that the code generated by the `addressForm` and `table` Schematics will fail a11y audits by default.

**Comment by papavlic** [comment]
Any update on a resolution for these issues? Are there any known workarounds for the time being until this is resolved?

**Comment by jrood** [comment]
@papavlic I'm not sure if this resolves your need, but in version 7 you can use a native select.

**Comment by papavlic** [comment]
Unfortunately we are somewhat heavily invested in mat-select and some features that come with it. Switching to native select would force us to do some major refactoring so we were just wondering when (if) this issue would get resolved as it would be great to have these accessibility features built in.

**Comment by jelbourn** [comment]
@crisbeto has been looking into it. The initial attempt at fixing this broke a huge number of unit tests inside Google, so we're trying another approach that will be more opt-in.

**Comment by papavlic** [comment]
@crisbeto do you have a general ETA for when this would land? Just wondering if its something that could be solved in a few months or is it a 6 months+ timeframe?

**Comment by kammokiran** [comment]
Hi, I have used Arya label for The select element to include the existing label plus the currently selected option. Four example if the label off the select box is‘ State‘, and if I have selected the option as‘ Karnataka‘, the aria label should be updated as‘ State Karnataka‘.

**Comment by boriskozak** [comment]
@crisbeto @jelbourn do you happen to have an update?

**Comment by Splaktar** [comment]
@boriskozak the only update is that this got looked at again in December and January via a new PR (https://github.com/angular/components/pull/14430) with a different approach. It doesn't look like it will make it into version `8.0.0`. Beyond that I'm not sure, but I wouldn't expect it before July at this point and it could be later than that.

**Comment by zuhairmahd** [comment]
Sorry to hear this won't make it into 8.  I'm hoping we would see a resolution sometime soon.

**Comment by numsu** [comment]
Just dropping a comment that we're also waiting for this. It's soon legally required to comply with WCAG standard a11y in our projects and we're also invested in using mat-select, would feel bad to fallback to native. It's the only component still failing.

**Comment by jelbourn** [comment]
We're now planning to address the issues as part of integrating the MDC Web components into Angular Material. The earlier attempts at fixing this were all just too much of a breaking change, such that it was impossible to roll out to applications in Google already using `MatSelect`.

**Comment by igoro1975** [comment]
Until the bug is fixed what are the other options to replace the control? Native is not approved by the business: "looks bad".
Thank you.

**Comment by Splaktar** [comment]
@igoro1975 you could possibly look at using directly or wrapping [`mdc-select`](https://material-components.github.io/material-components-web-catalog/#/component/select). There are some accessibility tips in the [MDC Select Docs](https://material.io/develop/web/components/input-controls/select-menus/).

**Comment by Splaktar** [comment]
Note that using the "Enhanced" (with Material Design popup panel) `mdc-select` version as a workaround for a11y issues, might not work. It uses the WAI-ARIA [Collapsible Dropdown Listbox Example](https://www.w3.org/TR/wai-aria-practices/examples/listbox/listbox-collapsible.html) which has some issues with VoiceOver and NVDA.

Another option is the `mdc-select__native-control` (with a native popup panel), but you can already get similar functionality from a `mat-form-field` that contains a native `select`. [Our docs](https://material.angular.io/components/select/overview) have many examples of this. This still gets you a Material Design form element, but the popup will be native (i.e. better a11y and better UX on iOS/Android).

Hopefully MDC Select gets some a11y improvements over the next few months, before [material-experimental](https://github.com/angular/components/tree/master/src/material-experimental) has an `mdc-select` implementation.

**Comment by zunaid2998** [comment]
Is there any fix yet on this issue ?

**Comment by 412andrewmortimer** [comment]
For what it's worth:

I arrived here after our A11y scanner library threw this error in e2e.

```
Fail:  aXe - Certain ARIA roles must contain particular children

<mat-select OMITTED...>

    https://dequeuniversity.com/rules/axe/3.4/aria-required-children?application=webdriverjs
```

And let's just say hopping around these issues has been confusing. But simply putting `role="listbox"` on my `<mat-select>`  and `role="option"` on my `<mat-option> ` fixed my error.

```
<!-- Like this -->
<mat-select role="listbox">
  <mat-option role="option"></mat-option>
</mat-select>
```

Am I not understanding something about the bigger picture or how A11y works in general?

**Comment by leomendoza123** [comment]
Is there any plan to solve this issue soon?

**Comment by Splaktar** [comment]
@412andrewmortimer that might trick the scanner, but to verify if that helped or hurt (possible), you would need to test your app with screen readers like JAWS/NVDA (Windows), ChromeVox (ChromeOS), Talkback (Android), and VoiceOver (macOS/iOS).

**Comment by Splaktar** [comment]
Some updates for my previous [`mdc-select`](https://material-components.github.io/material-components-web-catalog/#/component/select) mentions
- In `4.0.0`, `material-components-web` removed support for native `select`. The previously named "Enhanced" select is now the only option.
- That same update broke a number of form-related use cases because it also removed the hidden `input` element. There is a PR (https://github.com/material-components/material-components-web/issues/5428) to bring this back, but when it will be merged is unclear.
- This `4.0.0` update is supposed to have improved the a11y and functionality of `mdc-select`, but I have not yet tested this in a screen reader.

There has been some thinking about the MDC-related `mat-select` in our `material-experiemental` package, but there is no implementation ready to try yet and no ETA is available. You can watch our [material-experimental/mdc-select directory](https://github.com/angular/components/tree/master/src/material-experimental/mdc-select) or the section after "Contributing" in the [README](https://github.com/angular/components/blob/master/README.md) for updated plans each quarter.

**Comment by dorival** [comment]
I have an a11y issue related to JAWS 2019.1 and `MatInput`.

**Behaviour:** Multiple MatInputs in the form causes JAWS to speak the previous MatInput value as it jumps to the next MatInput field.

@Splaktar should I create a different issue, or do you prefer tracking it in here? I need a north so I can contribute towards the fix of this issue in particular.

**Comment by Splaktar** [comment]
@dorival please open a new issue for MatInput. Thank you.

**Comment by mkslalom** [comment]
Hi is there any chance this bug will get fixed soon? :) Please and thank you.

**Comment by Splaktar** [comment]
> Hi is there any chance this bug will get fixed soon? :) Please and thank you.

Based on https://github.com/angular/components/issues/11083#issuecomment-506541838, it's unlikely.
 
> Is there a workaround?

Not at this time. The new MDC-based version of `MatSelect` should help with this, but it hasn't been implemented yet.

**Comment by Josee9988** [comment]
When we will be able to use it without that accessibility issue?
